New model run 7th march

In [1]:
import torch
import torch.nn as nn
import torch.nn.functional as F

class SEBlock(nn.Module):
    def __init__(self, in_channels, reduction=16):
        super(SEBlock, self).__init__()
        self.fc1 = nn.Linear(in_channels, in_channels // reduction)
        self.fc2 = nn.Linear(in_channels // reduction, in_channels)

    def forward(self, x):
        batch, channels, _, _ = x.size()
        y = x.mean(dim=[2, 3])
        y = F.relu(self.fc1(y))
        y = torch.sigmoid(self.fc2(y)).view(batch, channels, 1, 1)
        return x * y

class BottleneckBlock(nn.Module):
    def __init__(self, in_planes, planes, stride=1, expansion=4):
        super(BottleneckBlock, self).__init__()
        self.conv1 = nn.Conv2d(in_planes, planes, kernel_size=1, bias=False)
        self.bn1 = nn.GroupNorm(4, planes)
        self.conv2 = nn.Conv2d(planes, planes, kernel_size=3, stride=stride, padding=1, groups=planes, bias=False)
        self.bn2 = nn.GroupNorm(4, planes)
        self.conv3 = nn.Conv2d(planes, planes * expansion, kernel_size=1, bias=False)
        self.bn3 = nn.GroupNorm(4, planes * expansion)
        self.se = SEBlock(planes * expansion)
        self.shortcut = nn.Sequential()
        if stride != 1 or in_planes != planes * expansion:
            self.shortcut = nn.Sequential(
                nn.Conv2d(in_planes, planes * expansion, kernel_size=1, stride=stride, bias=False),
                nn.GroupNorm(4, planes * expansion)
            )

    def forward(self, x):
        out = F.relu(self.bn1(self.conv1(x)))
        out = F.relu(self.bn2(self.conv2(out)))
        out = self.bn3(self.conv3(out))
        out = self.se(out)
        out += self.shortcut(x)
        return F.relu(out)

class ResNet(nn.Module):
    def __init__(self, block, num_blocks, num_classes=10):
        super(ResNet, self).__init__()
        self.in_planes = 64
        self.conv1 = nn.Conv2d(3, 64, kernel_size=3, stride=1, padding=1, bias=False)
        self.bn1 = nn.GroupNorm(4, 64)
        self.layer1 = self._make_layer(block, 64, num_blocks[0], stride=1)
        self.layer2 = self._make_layer(block, 128, num_blocks[1], stride=2)
        self.layer3 = self._make_layer(block, 256, num_blocks[2], stride=2)
        self.layer4 = self._make_layer(block, 512, num_blocks[3], stride=2)
        self.avgpool = nn.AdaptiveAvgPool2d(1)
        self.fc = nn.Linear(512 * 4, num_classes)

    def _make_layer(self, block, planes, num_blocks, stride):
        strides = [stride] + [1] * (num_blocks - 1)
        layers = []
        for stride in strides:
            layers.append(block(self.in_planes, planes, stride))
            self.in_planes = planes * 4
        return nn.Sequential(*layers)

    def forward(self, x):
        out = F.relu(self.bn1(self.conv1(x)))
        out = self.layer1(out)
        out = self.layer2(out)
        out = self.layer3(out)
        out = self.layer4(out)
        out = self.avgpool(out)
        out = torch.flatten(out, 1)
        return self.fc(out)

def ResNet18CIFAR():
    return ResNet(BottleneckBlock, [2, 2, 2, 2])

In [3]:
import torch
import torch.nn as nn
import torch.optim as optim
import os
import pickle
import numpy as np
from torch.utils.data import Dataset, DataLoader
import torchvision.transforms as transforms
from tqdm import tqdm
#from model import ResNet18CIFAR
import timm.optim.lookahead as Lookahead
from timm.optim.lookahead import Lookahead
# Custom Dataset Loader for CIFAR-10
class CIFARDataset(Dataset):
    def __init__(self, root_dir, train=True, transform=None):
        self.root_dir = root_dir
        self.transform = transform
        self.data, self.labels = self._load_data(train)

    def _load_data(self, train):
        data_list, labels_list = [], []
        batches = [f"data_batch_{i}" for i in range(1, 6)] if train else ["test_batch"]
        for batch in batches:
            batch_path = os.path.join(self.root_dir, batch)
            with open(batch_path, 'rb') as f:
                batch_dict = pickle.load(f, encoding='bytes')
                data_list.append(batch_dict[b'data'])
                labels_list.extend(batch_dict[b'labels'])
        X = np.concatenate(data_list, axis=0)
        y = np.array(labels_list)
        return X, y

    def __len__(self):
        return len(self.data)

    def __getitem__(self, idx):
        img = self.data[idx].reshape(3, 32, 32).transpose(1, 2, 0).astype("uint8")
        label = self.labels[idx]
        if self.transform:
            img = self.transform(img)
        return img, label

def train():
    device = 'cuda' if torch.cuda.is_available() else 'mps' if torch.backends.mps.is_available() else 'cpu'
    print(f"Using device: {device}")

    transform_train = transforms.Compose([
        transforms.ToPILImage(),
        transforms.RandomCrop(32, padding=4),
        transforms.RandomHorizontalFlip(),
        transforms.AutoAugment(),
        transforms.ToTensor(),
        transforms.Normalize((0.4914, 0.4822, 0.4465), (0.2470, 0.2435, 0.2616))
    ])
    transform_val = transforms.Compose([
        transforms.ToPILImage(),
        transforms.ToTensor(),
        transforms.Normalize((0.4914, 0.4822, 0.4465), (0.2470, 0.2435, 0.2616))
    ])

    root_dir = "/content/cifar-10-python/cifar-10-batches-py"
    trainset = CIFARDataset(root_dir=root_dir, train=True, transform=transform_train)
    valset = CIFARDataset(root_dir=root_dir, train=False, transform=transform_val)

    trainloader = DataLoader(trainset, batch_size=512, shuffle=True, num_workers=4)
    valloader = DataLoader(valset, batch_size=256, shuffle=False, num_workers=4)

    model = ResNet18CIFAR().to(device)
    criterion = nn.CrossEntropyLoss(label_smoothing=0.1)
    base_optimizer = optim.AdamW(model.parameters(), lr=3e-4, weight_decay=5e-4)
    optimizer = Lookahead(base_optimizer)  # Correct usage
    scheduler = optim.lr_scheduler.CosineAnnealingWarmRestarts(optimizer, T_0=10, T_mult=2)

    scaler = torch.cuda.amp.GradScaler()
    best_val_acc = 0.0

    for epoch in range(300):
        model.train()
        running_loss = 0.0
        for inputs, labels in tqdm(trainloader, desc=f"Epoch {epoch+1}/300"):
            inputs, labels = inputs.to(device), labels.to(device)
            optimizer.zero_grad()

            with torch.cuda.amp.autocast():
                outputs = model(inputs)
                loss = criterion(outputs, labels)

            scaler.scale(loss).backward()
            scaler.step(optimizer)
            scaler.update()
            running_loss += loss.item()
        scheduler.step()

        model.eval()
        correct = 0
        total = 0
        with torch.no_grad():
            for inputs, labels in valloader:
                inputs, labels = inputs.to(device), labels.to(device)
                outputs = model(inputs)
                _, predicted = torch.max(outputs, 1)
                total += labels.size(0)
                correct += (predicted == labels).sum().item()
        val_acc = 100.0 * correct / total
        print(f"Epoch [{epoch+1}/300] Loss: {running_loss/len(trainloader):.4f}, Val Acc: {val_acc:.2f}%")

        if val_acc > best_val_acc:
            best_val_acc = val_acc
            torch.save(model.state_dict(), "best_model.pth")
            print(f"Saved best model at epoch {epoch+1} with Val Acc: {val_acc:.2f}%")
    print("Training complete. Best Accuracy: {:.2f}%".format(best_val_acc))

if __name__ == "__main__":
    train()

Using device: cuda


<ipython-input-3-5db46cb0af60>:74: FutureWarning: `torch.cuda.amp.GradScaler(args...)` is deprecated. Please use `torch.amp.GradScaler('cuda', args...)` instead.
  scaler = torch.cuda.amp.GradScaler()
Epoch 1/300:   0%|          | 0/98 [00:00<?, ?it/s]<ipython-input-3-5db46cb0af60>:84: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.
  with torch.cuda.amp.autocast():
Epoch 1/300: 100%|██████████| 98/98 [00:16<00:00,  5.78it/s]


Epoch [1/300] Loss: 2.2517, Val Acc: 29.66%
Saved best model at epoch 1 with Val Acc: 29.66%


Epoch 2/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [2/300] Loss: 2.0606, Val Acc: 38.30%
Saved best model at epoch 2 with Val Acc: 38.30%


Epoch 3/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [3/300] Loss: 1.9221, Val Acc: 45.68%
Saved best model at epoch 3 with Val Acc: 45.68%


Epoch 4/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [4/300] Loss: 1.8091, Val Acc: 50.89%
Saved best model at epoch 4 with Val Acc: 50.89%


Epoch 5/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [5/300] Loss: 1.7243, Val Acc: 54.03%
Saved best model at epoch 5 with Val Acc: 54.03%


Epoch 6/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [6/300] Loss: 1.6466, Val Acc: 57.95%
Saved best model at epoch 6 with Val Acc: 57.95%


Epoch 7/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [7/300] Loss: 1.5809, Val Acc: 59.45%
Saved best model at epoch 7 with Val Acc: 59.45%


Epoch 8/300: 100%|██████████| 98/98 [00:14<00:00,  6.87it/s]


Epoch [8/300] Loss: 1.5418, Val Acc: 60.45%
Saved best model at epoch 8 with Val Acc: 60.45%


Epoch 9/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [9/300] Loss: 1.5157, Val Acc: 61.71%
Saved best model at epoch 9 with Val Acc: 61.71%


Epoch 10/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [10/300] Loss: 1.5062, Val Acc: 62.20%
Saved best model at epoch 10 with Val Acc: 62.20%


Epoch 11/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [11/300] Loss: 1.6782, Val Acc: 60.59%


Epoch 12/300: 100%|██████████| 98/98 [00:14<00:00,  6.87it/s]


Epoch [12/300] Loss: 1.5376, Val Acc: 61.42%


Epoch 13/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [13/300] Loss: 1.4917, Val Acc: 65.47%
Saved best model at epoch 13 with Val Acc: 65.47%


Epoch 14/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [14/300] Loss: 1.4469, Val Acc: 66.46%
Saved best model at epoch 14 with Val Acc: 66.46%


Epoch 15/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [15/300] Loss: 1.4164, Val Acc: 65.48%


Epoch 16/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [16/300] Loss: 1.3964, Val Acc: 66.88%
Saved best model at epoch 16 with Val Acc: 66.88%


Epoch 17/300: 100%|██████████| 98/98 [00:14<00:00,  6.87it/s]


Epoch [17/300] Loss: 1.3597, Val Acc: 69.91%
Saved best model at epoch 17 with Val Acc: 69.91%


Epoch 18/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [18/300] Loss: 1.3192, Val Acc: 71.24%
Saved best model at epoch 18 with Val Acc: 71.24%


Epoch 19/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [19/300] Loss: 1.2915, Val Acc: 72.38%
Saved best model at epoch 19 with Val Acc: 72.38%


Epoch 20/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [20/300] Loss: 1.2651, Val Acc: 73.22%
Saved best model at epoch 20 with Val Acc: 73.22%


Epoch 21/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [21/300] Loss: 1.2548, Val Acc: 74.01%
Saved best model at epoch 21 with Val Acc: 74.01%


Epoch 22/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [22/300] Loss: 1.2275, Val Acc: 74.86%
Saved best model at epoch 22 with Val Acc: 74.86%


Epoch 23/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [23/300] Loss: 1.2139, Val Acc: 74.08%


Epoch 24/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [24/300] Loss: 1.1967, Val Acc: 74.74%


Epoch 25/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [25/300] Loss: 1.1819, Val Acc: 75.88%
Saved best model at epoch 25 with Val Acc: 75.88%


Epoch 26/300: 100%|██████████| 98/98 [00:14<00:00,  6.87it/s]


Epoch [26/300] Loss: 1.1721, Val Acc: 75.17%


Epoch 27/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [27/300] Loss: 1.1663, Val Acc: 76.31%
Saved best model at epoch 27 with Val Acc: 76.31%


Epoch 28/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [28/300] Loss: 1.1616, Val Acc: 76.78%
Saved best model at epoch 28 with Val Acc: 76.78%


Epoch 29/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [29/300] Loss: 1.1553, Val Acc: 76.81%
Saved best model at epoch 29 with Val Acc: 76.81%


Epoch 30/300: 100%|██████████| 98/98 [00:14<00:00,  6.87it/s]


Epoch [30/300] Loss: 1.1543, Val Acc: 76.85%
Saved best model at epoch 30 with Val Acc: 76.85%


Epoch 31/300: 100%|██████████| 98/98 [00:14<00:00,  6.87it/s]


Epoch [31/300] Loss: 1.3174, Val Acc: 75.27%


Epoch 32/300: 100%|██████████| 98/98 [00:14<00:00,  6.87it/s]


Epoch [32/300] Loss: 1.2296, Val Acc: 74.54%


Epoch 33/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [33/300] Loss: 1.2051, Val Acc: 75.09%


Epoch 34/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [34/300] Loss: 1.1857, Val Acc: 76.43%


Epoch 35/300: 100%|██████████| 98/98 [00:14<00:00,  6.87it/s]


Epoch [35/300] Loss: 1.1729, Val Acc: 76.91%
Saved best model at epoch 35 with Val Acc: 76.91%


Epoch 36/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [36/300] Loss: 1.1556, Val Acc: 75.72%


Epoch 37/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [37/300] Loss: 1.1392, Val Acc: 77.99%
Saved best model at epoch 37 with Val Acc: 77.99%


Epoch 38/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [38/300] Loss: 1.1307, Val Acc: 77.47%


Epoch 39/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [39/300] Loss: 1.1152, Val Acc: 78.69%
Saved best model at epoch 39 with Val Acc: 78.69%


Epoch 40/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [40/300] Loss: 1.1029, Val Acc: 79.10%
Saved best model at epoch 40 with Val Acc: 79.10%


Epoch 41/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [41/300] Loss: 1.0873, Val Acc: 79.87%
Saved best model at epoch 41 with Val Acc: 79.87%


Epoch 42/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [42/300] Loss: 1.0782, Val Acc: 81.12%
Saved best model at epoch 42 with Val Acc: 81.12%


Epoch 43/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [43/300] Loss: 1.0734, Val Acc: 80.21%


Epoch 44/300: 100%|██████████| 98/98 [00:14<00:00,  6.87it/s]


Epoch [44/300] Loss: 1.0604, Val Acc: 80.98%


Epoch 45/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [45/300] Loss: 1.0493, Val Acc: 80.74%


Epoch 46/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [46/300] Loss: 1.0388, Val Acc: 82.34%
Saved best model at epoch 46 with Val Acc: 82.34%


Epoch 47/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [47/300] Loss: 1.0327, Val Acc: 81.52%


Epoch 48/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [48/300] Loss: 1.0210, Val Acc: 80.17%


Epoch 49/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [49/300] Loss: 1.0105, Val Acc: 81.41%


Epoch 50/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [50/300] Loss: 1.0083, Val Acc: 81.98%


Epoch 51/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [51/300] Loss: 0.9960, Val Acc: 81.01%


Epoch 52/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [52/300] Loss: 0.9923, Val Acc: 82.96%
Saved best model at epoch 52 with Val Acc: 82.96%


Epoch 53/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [53/300] Loss: 0.9804, Val Acc: 82.69%


Epoch 54/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [54/300] Loss: 0.9793, Val Acc: 81.86%


Epoch 55/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [55/300] Loss: 0.9680, Val Acc: 83.32%
Saved best model at epoch 55 with Val Acc: 83.32%


Epoch 56/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [56/300] Loss: 0.9658, Val Acc: 83.15%


Epoch 57/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [57/300] Loss: 0.9571, Val Acc: 83.01%


Epoch 58/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [58/300] Loss: 0.9519, Val Acc: 83.71%
Saved best model at epoch 58 with Val Acc: 83.71%


Epoch 59/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [59/300] Loss: 0.9529, Val Acc: 83.83%
Saved best model at epoch 59 with Val Acc: 83.83%


Epoch 60/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [60/300] Loss: 0.9456, Val Acc: 83.94%
Saved best model at epoch 60 with Val Acc: 83.94%


Epoch 61/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [61/300] Loss: 0.9379, Val Acc: 83.93%


Epoch 62/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [62/300] Loss: 0.9316, Val Acc: 84.25%
Saved best model at epoch 62 with Val Acc: 84.25%


Epoch 63/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [63/300] Loss: 0.9354, Val Acc: 84.22%


Epoch 64/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [64/300] Loss: 0.9295, Val Acc: 84.25%


Epoch 65/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [65/300] Loss: 0.9275, Val Acc: 84.46%
Saved best model at epoch 65 with Val Acc: 84.46%


Epoch 66/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [66/300] Loss: 0.9227, Val Acc: 84.13%


Epoch 67/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [67/300] Loss: 0.9242, Val Acc: 84.22%


Epoch 68/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [68/300] Loss: 0.9240, Val Acc: 84.27%


Epoch 69/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [69/300] Loss: 0.9238, Val Acc: 84.27%


Epoch 70/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [70/300] Loss: 0.9201, Val Acc: 84.37%


Epoch 71/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [71/300] Loss: 1.0537, Val Acc: 81.78%


Epoch 72/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [72/300] Loss: 1.0027, Val Acc: 81.33%


Epoch 73/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [73/300] Loss: 0.9968, Val Acc: 83.09%


Epoch 74/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [74/300] Loss: 0.9904, Val Acc: 83.67%


Epoch 75/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [75/300] Loss: 0.9933, Val Acc: 80.99%


Epoch 76/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [76/300] Loss: 0.9870, Val Acc: 83.04%


Epoch 77/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [77/300] Loss: 0.9803, Val Acc: 82.53%


Epoch 78/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [78/300] Loss: 0.9673, Val Acc: 83.84%


Epoch 79/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [79/300] Loss: 0.9648, Val Acc: 83.81%


Epoch 80/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [80/300] Loss: 0.9569, Val Acc: 82.83%


Epoch 81/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [81/300] Loss: 0.9535, Val Acc: 84.36%


Epoch 82/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [82/300] Loss: 0.9424, Val Acc: 84.49%
Saved best model at epoch 82 with Val Acc: 84.49%


Epoch 83/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [83/300] Loss: 0.9408, Val Acc: 82.94%


Epoch 84/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [84/300] Loss: 0.9274, Val Acc: 84.72%
Saved best model at epoch 84 with Val Acc: 84.72%


Epoch 85/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [85/300] Loss: 0.9309, Val Acc: 84.84%
Saved best model at epoch 85 with Val Acc: 84.84%


Epoch 86/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [86/300] Loss: 0.9236, Val Acc: 83.91%


Epoch 87/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [87/300] Loss: 0.9271, Val Acc: 84.23%


Epoch 88/300: 100%|██████████| 98/98 [00:14<00:00,  6.82it/s]


Epoch [88/300] Loss: 0.9163, Val Acc: 85.33%
Saved best model at epoch 88 with Val Acc: 85.33%


Epoch 89/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [89/300] Loss: 0.9140, Val Acc: 84.54%


Epoch 90/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [90/300] Loss: 0.9048, Val Acc: 84.73%


Epoch 91/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [91/300] Loss: 0.8978, Val Acc: 85.88%
Saved best model at epoch 91 with Val Acc: 85.88%


Epoch 92/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [92/300] Loss: 0.8971, Val Acc: 85.61%


Epoch 93/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [93/300] Loss: 0.8953, Val Acc: 84.36%


Epoch 94/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [94/300] Loss: 0.8938, Val Acc: 85.96%
Saved best model at epoch 94 with Val Acc: 85.96%


Epoch 95/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [95/300] Loss: 0.8877, Val Acc: 85.70%


Epoch 96/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [96/300] Loss: 0.8762, Val Acc: 85.58%


Epoch 97/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [97/300] Loss: 0.8818, Val Acc: 86.18%
Saved best model at epoch 97 with Val Acc: 86.18%


Epoch 98/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [98/300] Loss: 0.8637, Val Acc: 84.87%


Epoch 99/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [99/300] Loss: 0.8669, Val Acc: 86.35%
Saved best model at epoch 99 with Val Acc: 86.35%


Epoch 100/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [100/300] Loss: 0.8636, Val Acc: 86.89%
Saved best model at epoch 100 with Val Acc: 86.89%


Epoch 101/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [101/300] Loss: 0.8542, Val Acc: 85.93%


Epoch 102/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [102/300] Loss: 0.8542, Val Acc: 86.08%


Epoch 103/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [103/300] Loss: 0.8449, Val Acc: 86.23%


Epoch 104/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [104/300] Loss: 0.8509, Val Acc: 86.21%


Epoch 105/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [105/300] Loss: 0.8415, Val Acc: 86.66%


Epoch 106/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [106/300] Loss: 0.8420, Val Acc: 86.85%


Epoch 107/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [107/300] Loss: 0.8330, Val Acc: 86.71%


Epoch 108/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [108/300] Loss: 0.8329, Val Acc: 87.03%
Saved best model at epoch 108 with Val Acc: 87.03%


Epoch 109/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [109/300] Loss: 0.8249, Val Acc: 87.51%
Saved best model at epoch 109 with Val Acc: 87.51%


Epoch 110/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [110/300] Loss: 0.8268, Val Acc: 86.60%


Epoch 111/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [111/300] Loss: 0.8244, Val Acc: 87.36%


Epoch 112/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [112/300] Loss: 0.8132, Val Acc: 87.53%
Saved best model at epoch 112 with Val Acc: 87.53%


Epoch 113/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [113/300] Loss: 0.8164, Val Acc: 87.28%


Epoch 114/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [114/300] Loss: 0.8070, Val Acc: 86.79%


Epoch 115/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [115/300] Loss: 0.8069, Val Acc: 87.53%


Epoch 116/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [116/300] Loss: 0.7983, Val Acc: 87.94%
Saved best model at epoch 116 with Val Acc: 87.94%


Epoch 117/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [117/300] Loss: 0.8010, Val Acc: 87.60%


Epoch 118/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [118/300] Loss: 0.7960, Val Acc: 87.93%


Epoch 119/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [119/300] Loss: 0.7947, Val Acc: 87.69%


Epoch 120/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [120/300] Loss: 0.7910, Val Acc: 87.40%


Epoch 121/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [121/300] Loss: 0.7879, Val Acc: 88.17%
Saved best model at epoch 121 with Val Acc: 88.17%


Epoch 122/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [122/300] Loss: 0.7820, Val Acc: 87.82%


Epoch 123/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [123/300] Loss: 0.7833, Val Acc: 88.01%


Epoch 124/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [124/300] Loss: 0.7792, Val Acc: 88.14%


Epoch 125/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [125/300] Loss: 0.7760, Val Acc: 87.58%


Epoch 126/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [126/300] Loss: 0.7722, Val Acc: 88.45%
Saved best model at epoch 126 with Val Acc: 88.45%


Epoch 127/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [127/300] Loss: 0.7699, Val Acc: 88.36%


Epoch 128/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [128/300] Loss: 0.7715, Val Acc: 88.13%


Epoch 129/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [129/300] Loss: 0.7655, Val Acc: 88.36%


Epoch 130/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [130/300] Loss: 0.7628, Val Acc: 88.50%
Saved best model at epoch 130 with Val Acc: 88.50%


Epoch 131/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [131/300] Loss: 0.7623, Val Acc: 88.41%


Epoch 132/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [132/300] Loss: 0.7595, Val Acc: 88.38%


Epoch 133/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [133/300] Loss: 0.7618, Val Acc: 88.26%


Epoch 134/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [134/300] Loss: 0.7559, Val Acc: 88.63%
Saved best model at epoch 134 with Val Acc: 88.63%


Epoch 135/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [135/300] Loss: 0.7564, Val Acc: 88.36%


Epoch 136/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [136/300] Loss: 0.7526, Val Acc: 88.34%


Epoch 137/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [137/300] Loss: 0.7547, Val Acc: 88.38%


Epoch 138/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [138/300] Loss: 0.7521, Val Acc: 88.48%


Epoch 139/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [139/300] Loss: 0.7534, Val Acc: 88.31%


Epoch 140/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [140/300] Loss: 0.7492, Val Acc: 88.32%


Epoch 141/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [141/300] Loss: 0.7468, Val Acc: 88.42%


Epoch 142/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [142/300] Loss: 0.7526, Val Acc: 88.35%


Epoch 143/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [143/300] Loss: 0.7458, Val Acc: 88.41%


Epoch 144/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [144/300] Loss: 0.7457, Val Acc: 88.40%


Epoch 145/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [145/300] Loss: 0.7479, Val Acc: 88.26%


Epoch 146/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [146/300] Loss: 0.7462, Val Acc: 88.41%


Epoch 147/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [147/300] Loss: 0.7475, Val Acc: 88.44%


Epoch 148/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [148/300] Loss: 0.7448, Val Acc: 88.47%


Epoch 149/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [149/300] Loss: 0.7467, Val Acc: 88.49%


Epoch 150/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [150/300] Loss: 0.7484, Val Acc: 88.45%


Epoch 151/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [151/300] Loss: 0.8496, Val Acc: 85.95%


Epoch 152/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [152/300] Loss: 0.8393, Val Acc: 87.32%


Epoch 153/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [153/300] Loss: 0.8462, Val Acc: 86.51%


Epoch 154/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [154/300] Loss: 0.8324, Val Acc: 86.44%


Epoch 155/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [155/300] Loss: 0.8383, Val Acc: 87.46%


Epoch 156/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [156/300] Loss: 0.8247, Val Acc: 86.72%


Epoch 157/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [157/300] Loss: 0.8225, Val Acc: 87.63%


Epoch 158/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [158/300] Loss: 0.8204, Val Acc: 87.88%


Epoch 159/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [159/300] Loss: 0.8133, Val Acc: 87.86%


Epoch 160/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [160/300] Loss: 0.8150, Val Acc: 87.12%


Epoch 161/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [161/300] Loss: 0.8071, Val Acc: 87.45%


Epoch 162/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [162/300] Loss: 0.8096, Val Acc: 87.34%


Epoch 163/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [163/300] Loss: 0.7981, Val Acc: 86.86%


Epoch 164/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [164/300] Loss: 0.7915, Val Acc: 87.36%


Epoch 165/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [165/300] Loss: 0.7977, Val Acc: 87.87%


Epoch 166/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [166/300] Loss: 0.7968, Val Acc: 87.25%


Epoch 167/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [167/300] Loss: 0.7937, Val Acc: 87.88%


Epoch 168/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [168/300] Loss: 0.7942, Val Acc: 86.94%


Epoch 169/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [169/300] Loss: 0.7935, Val Acc: 87.10%


Epoch 170/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [170/300] Loss: 0.7849, Val Acc: 88.24%


Epoch 171/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [171/300] Loss: 0.7831, Val Acc: 86.93%


Epoch 172/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [172/300] Loss: 0.7847, Val Acc: 86.70%


Epoch 173/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [173/300] Loss: 0.7809, Val Acc: 88.42%


Epoch 174/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [174/300] Loss: 0.7743, Val Acc: 88.54%


Epoch 175/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [175/300] Loss: 0.7690, Val Acc: 87.93%


Epoch 176/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [176/300] Loss: 0.7682, Val Acc: 88.32%


Epoch 177/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [177/300] Loss: 0.7690, Val Acc: 88.69%
Saved best model at epoch 177 with Val Acc: 88.69%


Epoch 178/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [178/300] Loss: 0.7659, Val Acc: 87.53%


Epoch 179/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [179/300] Loss: 0.7642, Val Acc: 88.88%
Saved best model at epoch 179 with Val Acc: 88.88%


Epoch 180/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [180/300] Loss: 0.7667, Val Acc: 89.08%
Saved best model at epoch 180 with Val Acc: 89.08%


Epoch 181/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [181/300] Loss: 0.7579, Val Acc: 88.25%


Epoch 182/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [182/300] Loss: 0.7570, Val Acc: 88.05%


Epoch 183/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [183/300] Loss: 0.7524, Val Acc: 88.50%


Epoch 184/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [184/300] Loss: 0.7565, Val Acc: 88.17%


Epoch 185/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [185/300] Loss: 0.7543, Val Acc: 87.84%


Epoch 186/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [186/300] Loss: 0.7496, Val Acc: 88.75%


Epoch 187/300: 100%|██████████| 98/98 [00:14<00:00,  6.82it/s]


Epoch [187/300] Loss: 0.7478, Val Acc: 88.56%


Epoch 188/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [188/300] Loss: 0.7468, Val Acc: 89.12%
Saved best model at epoch 188 with Val Acc: 89.12%


Epoch 189/300: 100%|██████████| 98/98 [00:14<00:00,  6.85it/s]


Epoch [189/300] Loss: 0.7431, Val Acc: 88.91%


Epoch 190/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [190/300] Loss: 0.7397, Val Acc: 88.94%


Epoch 191/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [191/300] Loss: 0.7399, Val Acc: 88.47%


Epoch 192/300: 100%|██████████| 98/98 [00:14<00:00,  6.86it/s]


Epoch [192/300] Loss: 0.7370, Val Acc: 87.87%


Epoch 193/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [193/300] Loss: 0.7281, Val Acc: 87.69%


Epoch 194/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [194/300] Loss: 0.7343, Val Acc: 88.10%


Epoch 195/300: 100%|██████████| 98/98 [00:14<00:00,  6.82it/s]


Epoch [195/300] Loss: 0.7310, Val Acc: 89.17%
Saved best model at epoch 195 with Val Acc: 89.17%


Epoch 196/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [196/300] Loss: 0.7314, Val Acc: 88.56%


Epoch 197/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [197/300] Loss: 0.7303, Val Acc: 88.49%


Epoch 198/300: 100%|██████████| 98/98 [00:14<00:00,  6.82it/s]


Epoch [198/300] Loss: 0.7260, Val Acc: 89.37%
Saved best model at epoch 198 with Val Acc: 89.37%


Epoch 199/300: 100%|██████████| 98/98 [00:14<00:00,  6.82it/s]


Epoch [199/300] Loss: 0.7256, Val Acc: 88.91%


Epoch 200/300: 100%|██████████| 98/98 [00:14<00:00,  6.82it/s]


Epoch [200/300] Loss: 0.7234, Val Acc: 88.78%


Epoch 201/300: 100%|██████████| 98/98 [00:14<00:00,  6.82it/s]


Epoch [201/300] Loss: 0.7152, Val Acc: 89.26%


Epoch 202/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [202/300] Loss: 0.7182, Val Acc: 89.20%


Epoch 203/300: 100%|██████████| 98/98 [00:14<00:00,  6.82it/s]


Epoch [203/300] Loss: 0.7176, Val Acc: 89.00%


Epoch 204/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [204/300] Loss: 0.7106, Val Acc: 89.35%


Epoch 205/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [205/300] Loss: 0.7132, Val Acc: 88.64%


Epoch 206/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [206/300] Loss: 0.7093, Val Acc: 89.33%


Epoch 207/300: 100%|██████████| 98/98 [00:14<00:00,  6.82it/s]


Epoch [207/300] Loss: 0.7108, Val Acc: 89.57%
Saved best model at epoch 207 with Val Acc: 89.57%


Epoch 208/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [208/300] Loss: 0.7027, Val Acc: 89.32%


Epoch 209/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [209/300] Loss: 0.7040, Val Acc: 89.29%


Epoch 210/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [210/300] Loss: 0.7017, Val Acc: 89.78%
Saved best model at epoch 210 with Val Acc: 89.78%


Epoch 211/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [211/300] Loss: 0.7027, Val Acc: 88.95%


Epoch 212/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [212/300] Loss: 0.7004, Val Acc: 88.37%


Epoch 213/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [213/300] Loss: 0.6928, Val Acc: 89.52%


Epoch 214/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [214/300] Loss: 0.6936, Val Acc: 88.98%


Epoch 215/300: 100%|██████████| 98/98 [00:14<00:00,  6.75it/s]


Epoch [215/300] Loss: 0.6942, Val Acc: 88.29%


Epoch 216/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [216/300] Loss: 0.6854, Val Acc: 89.43%


Epoch 217/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [217/300] Loss: 0.6860, Val Acc: 89.29%


Epoch 218/300: 100%|██████████| 98/98 [00:14<00:00,  6.75it/s]


Epoch [218/300] Loss: 0.6831, Val Acc: 88.65%


Epoch 219/300: 100%|██████████| 98/98 [00:14<00:00,  6.67it/s]


Epoch [219/300] Loss: 0.6869, Val Acc: 89.38%


Epoch 220/300: 100%|██████████| 98/98 [00:14<00:00,  6.73it/s]


Epoch [220/300] Loss: 0.6814, Val Acc: 89.15%


Epoch 221/300: 100%|██████████| 98/98 [00:14<00:00,  6.75it/s]


Epoch [221/300] Loss: 0.6828, Val Acc: 88.95%


Epoch 222/300: 100%|██████████| 98/98 [00:14<00:00,  6.72it/s]


Epoch [222/300] Loss: 0.6824, Val Acc: 89.80%
Saved best model at epoch 222 with Val Acc: 89.80%


Epoch 223/300: 100%|██████████| 98/98 [00:14<00:00,  6.73it/s]


Epoch [223/300] Loss: 0.6769, Val Acc: 89.28%


Epoch 224/300: 100%|██████████| 98/98 [00:14<00:00,  6.73it/s]


Epoch [224/300] Loss: 0.6742, Val Acc: 89.53%


Epoch 225/300: 100%|██████████| 98/98 [00:14<00:00,  6.71it/s]


Epoch [225/300] Loss: 0.6754, Val Acc: 89.30%


Epoch 226/300: 100%|██████████| 98/98 [00:14<00:00,  6.74it/s]


Epoch [226/300] Loss: 0.6743, Val Acc: 89.27%


Epoch 227/300: 100%|██████████| 98/98 [00:14<00:00,  6.73it/s]


Epoch [227/300] Loss: 0.6714, Val Acc: 89.31%


Epoch 228/300: 100%|██████████| 98/98 [00:14<00:00,  6.71it/s]


Epoch [228/300] Loss: 0.6693, Val Acc: 89.65%


Epoch 229/300: 100%|██████████| 98/98 [00:14<00:00,  6.77it/s]


Epoch [229/300] Loss: 0.6653, Val Acc: 89.84%
Saved best model at epoch 229 with Val Acc: 89.84%


Epoch 230/300: 100%|██████████| 98/98 [00:14<00:00,  6.73it/s]


Epoch [230/300] Loss: 0.6680, Val Acc: 88.91%


Epoch 231/300: 100%|██████████| 98/98 [00:14<00:00,  6.79it/s]


Epoch [231/300] Loss: 0.6692, Val Acc: 89.76%


Epoch 232/300: 100%|██████████| 98/98 [00:14<00:00,  6.76it/s]


Epoch [232/300] Loss: 0.6598, Val Acc: 89.44%


Epoch 233/300: 100%|██████████| 98/98 [00:14<00:00,  6.74it/s]


Epoch [233/300] Loss: 0.6632, Val Acc: 89.08%


Epoch 234/300: 100%|██████████| 98/98 [00:14<00:00,  6.72it/s]


Epoch [234/300] Loss: 0.6592, Val Acc: 89.75%


Epoch 235/300: 100%|██████████| 98/98 [00:14<00:00,  6.74it/s]


Epoch [235/300] Loss: 0.6571, Val Acc: 90.08%
Saved best model at epoch 235 with Val Acc: 90.08%


Epoch 236/300: 100%|██████████| 98/98 [00:14<00:00,  6.70it/s]


Epoch [236/300] Loss: 0.6564, Val Acc: 89.63%


Epoch 237/300: 100%|██████████| 98/98 [00:14<00:00,  6.59it/s]


Epoch [237/300] Loss: 0.6571, Val Acc: 90.06%


Epoch 238/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [238/300] Loss: 0.6516, Val Acc: 89.34%


Epoch 239/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [239/300] Loss: 0.6536, Val Acc: 89.62%


Epoch 240/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [240/300] Loss: 0.6535, Val Acc: 90.08%


Epoch 241/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [241/300] Loss: 0.6478, Val Acc: 89.65%


Epoch 242/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [242/300] Loss: 0.6507, Val Acc: 89.37%


Epoch 243/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [243/300] Loss: 0.6474, Val Acc: 89.43%


Epoch 244/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [244/300] Loss: 0.6471, Val Acc: 89.86%


Epoch 245/300: 100%|██████████| 98/98 [00:14<00:00,  6.82it/s]


Epoch [245/300] Loss: 0.6451, Val Acc: 89.94%


Epoch 246/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [246/300] Loss: 0.6425, Val Acc: 90.07%


Epoch 247/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [247/300] Loss: 0.6409, Val Acc: 89.72%


Epoch 248/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [248/300] Loss: 0.6438, Val Acc: 89.86%


Epoch 249/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [249/300] Loss: 0.6372, Val Acc: 89.51%


Epoch 250/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [250/300] Loss: 0.6402, Val Acc: 89.79%


Epoch 251/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [251/300] Loss: 0.6397, Val Acc: 90.04%


Epoch 252/300: 100%|██████████| 98/98 [00:14<00:00,  6.82it/s]


Epoch [252/300] Loss: 0.6360, Val Acc: 90.07%


Epoch 253/300: 100%|██████████| 98/98 [00:14<00:00,  6.82it/s]


Epoch [253/300] Loss: 0.6347, Val Acc: 89.72%


Epoch 254/300: 100%|██████████| 98/98 [00:14<00:00,  6.82it/s]


Epoch [254/300] Loss: 0.6297, Val Acc: 89.47%


Epoch 255/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [255/300] Loss: 0.6337, Val Acc: 89.76%


Epoch 256/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [256/300] Loss: 0.6328, Val Acc: 90.04%


Epoch 257/300: 100%|██████████| 98/98 [00:14<00:00,  6.82it/s]


Epoch [257/300] Loss: 0.6286, Val Acc: 90.00%


Epoch 258/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [258/300] Loss: 0.6296, Val Acc: 90.05%


Epoch 259/300: 100%|██████████| 98/98 [00:14<00:00,  6.82it/s]


Epoch [259/300] Loss: 0.6251, Val Acc: 90.10%
Saved best model at epoch 259 with Val Acc: 90.10%


Epoch 260/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [260/300] Loss: 0.6273, Val Acc: 90.06%


Epoch 261/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [261/300] Loss: 0.6254, Val Acc: 89.72%


Epoch 262/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [262/300] Loss: 0.6238, Val Acc: 90.08%


Epoch 263/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [263/300] Loss: 0.6257, Val Acc: 90.07%


Epoch 264/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [264/300] Loss: 0.6259, Val Acc: 89.90%


Epoch 265/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [265/300] Loss: 0.6214, Val Acc: 90.59%
Saved best model at epoch 265 with Val Acc: 90.59%


Epoch 266/300: 100%|██████████| 98/98 [00:14<00:00,  6.82it/s]


Epoch [266/300] Loss: 0.6195, Val Acc: 90.06%


Epoch 267/300: 100%|██████████| 98/98 [00:14<00:00,  6.81it/s]


Epoch [267/300] Loss: 0.6196, Val Acc: 90.15%


Epoch 268/300: 100%|██████████| 98/98 [00:14<00:00,  6.82it/s]


Epoch [268/300] Loss: 0.6188, Val Acc: 89.91%


Epoch 269/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [269/300] Loss: 0.6154, Val Acc: 90.35%


Epoch 270/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [270/300] Loss: 0.6203, Val Acc: 90.01%


Epoch 271/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [271/300] Loss: 0.6188, Val Acc: 90.27%


Epoch 272/300: 100%|██████████| 98/98 [00:14<00:00,  6.82it/s]


Epoch [272/300] Loss: 0.6151, Val Acc: 90.05%


Epoch 273/300: 100%|██████████| 98/98 [00:14<00:00,  6.82it/s]


Epoch [273/300] Loss: 0.6142, Val Acc: 90.23%


Epoch 274/300: 100%|██████████| 98/98 [00:14<00:00,  6.82it/s]


Epoch [274/300] Loss: 0.6130, Val Acc: 90.04%


Epoch 275/300: 100%|██████████| 98/98 [00:14<00:00,  6.82it/s]


Epoch [275/300] Loss: 0.6098, Val Acc: 90.12%


Epoch 276/300: 100%|██████████| 98/98 [00:14<00:00,  6.82it/s]


Epoch [276/300] Loss: 0.6160, Val Acc: 90.33%


Epoch 277/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [277/300] Loss: 0.6121, Val Acc: 90.35%


Epoch 278/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [278/300] Loss: 0.6109, Val Acc: 90.31%


Epoch 279/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [279/300] Loss: 0.6102, Val Acc: 90.35%


Epoch 280/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [280/300] Loss: 0.6089, Val Acc: 90.33%


Epoch 281/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [281/300] Loss: 0.6130, Val Acc: 90.32%


Epoch 282/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [282/300] Loss: 0.6121, Val Acc: 90.55%


Epoch 283/300: 100%|██████████| 98/98 [00:14<00:00,  6.81it/s]


Epoch [283/300] Loss: 0.6073, Val Acc: 90.55%


Epoch 284/300: 100%|██████████| 98/98 [00:14<00:00,  6.82it/s]


Epoch [284/300] Loss: 0.6085, Val Acc: 90.41%


Epoch 285/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [285/300] Loss: 0.6075, Val Acc: 90.47%


Epoch 286/300: 100%|██████████| 98/98 [00:14<00:00,  6.82it/s]


Epoch [286/300] Loss: 0.6038, Val Acc: 90.35%


Epoch 287/300: 100%|██████████| 98/98 [00:14<00:00,  6.82it/s]


Epoch [287/300] Loss: 0.6055, Val Acc: 90.55%


Epoch 288/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [288/300] Loss: 0.6072, Val Acc: 90.31%


Epoch 289/300: 100%|██████████| 98/98 [00:14<00:00,  6.81it/s]


Epoch [289/300] Loss: 0.6051, Val Acc: 90.38%


Epoch 290/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [290/300] Loss: 0.6063, Val Acc: 90.39%


Epoch 291/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [291/300] Loss: 0.6032, Val Acc: 90.37%


Epoch 292/300: 100%|██████████| 98/98 [00:14<00:00,  6.82it/s]


Epoch [292/300] Loss: 0.6095, Val Acc: 90.40%


Epoch 293/300: 100%|██████████| 98/98 [00:14<00:00,  6.82it/s]


Epoch [293/300] Loss: 0.6071, Val Acc: 90.38%


Epoch 294/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [294/300] Loss: 0.6007, Val Acc: 90.44%


Epoch 295/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [295/300] Loss: 0.6052, Val Acc: 90.38%


Epoch 296/300: 100%|██████████| 98/98 [00:14<00:00,  6.84it/s]


Epoch [296/300] Loss: 0.6042, Val Acc: 90.40%


Epoch 297/300: 100%|██████████| 98/98 [00:14<00:00,  6.82it/s]


Epoch [297/300] Loss: 0.6037, Val Acc: 90.39%


Epoch 298/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [298/300] Loss: 0.6061, Val Acc: 90.35%


Epoch 299/300: 100%|██████████| 98/98 [00:14<00:00,  6.82it/s]


Epoch [299/300] Loss: 0.6038, Val Acc: 90.30%


Epoch 300/300: 100%|██████████| 98/98 [00:14<00:00,  6.83it/s]


Epoch [300/300] Loss: 0.6039, Val Acc: 90.29%
Training complete. Best Accuracy: 90.59%


In [12]:
import torch
import torchvision.transforms as transforms
import pandas as pd
import pickle
import os
import numpy as np
#from model import ResNet18CIFAR
from torch.utils.data import DataLoader, Dataset

def unpickle(file):
    with open(file, 'rb') as fo:
        data_dict = pickle.load(fo, encoding='bytes')
    return data_dict

class CIFARTestDataset(Dataset):
    def __init__(self, data, ids, transform=None):
        """
        data: numpy array of shape (N, 32, 32, 3)
        ids: array-like of test image IDs
        transform: transforms to apply
        """
        self.data = data
        self.ids = ids
        self.transform = transform

    def __len__(self):
        return len(self.data)

    def __getitem__(self, idx):
        img = self.data[idx].astype("uint8")
        img_id = self.ids[idx]
        if self.transform:
            img = self.transform(img)
        return img, img_id

def inference():
    device = 'cuda' if torch.cuda.is_available() else 'mps' if torch.backends.mps.is_available() else 'cpu'
    print(f"Using device: {device}")

    model = ResNet18CIFAR().to(device)
    model.load_state_dict(torch.load("best_model.pth", map_location=device))
    model.eval()

    test_file = "/content/cifar_test_nolabel.pkl"
    test_dict = unpickle(test_file)
    print("Keys in test_dict:", test_dict.keys())

    test_images = test_dict[b'data'].reshape(-1, 32, 32, 3)  # Ensure correct shape
    test_ids = [str(i) for i in range(len(test_images))]  # Generate sequential IDs

    transform_test = transforms.Compose([
        transforms.ToPILImage(),
        transforms.ToTensor(),
        transforms.Normalize((0.4914, 0.4822, 0.4465), (0.2470, 0.2435, 0.2616))
    ])

    test_dataset = CIFARTestDataset(test_images, test_ids, transform=transform_test)
    test_loader = DataLoader(test_dataset, batch_size=128, shuffle=False, num_workers=4)

    predictions = []
    image_ids = []

    with torch.no_grad():
        for images, ids in test_loader:
            images = images.to(device)
            outputs = model(images)
            probabilities = torch.nn.functional.softmax(outputs, dim=1)  # Apply Softmax
            _, predicted = torch.max(probabilities, 1)
            predictions.extend(predicted.cpu().numpy().tolist())
            image_ids.extend(ids)

    submission = pd.DataFrame({"ID": image_ids, "Labels": predictions})
    submission.to_csv("submission.csv", index=False)
    print("Submission file saved as submission.csv")

if __name__ == "__main__":
    inference()

Using device: cuda
Keys in test_dict: dict_keys([b'data', b'ids'])


<ipython-input-12-8547a9c001a3>:41: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  model.load_state_dict(torch.load("best_model.pth", map_location=device))


Submission file saved as submission.csv


This model and training fetched us 0.77 on kaggle.

In [23]:
import pandas as pd

# Load and inspect submission file
submission_df = pd.read_csv("submission.csv")
print(submission_df.head())  # Show first few rows
print(submission_df["Labels"].value_counts())  # Show label distribution


   ID  Labels
0   0       6
1   1       1
2   2       8
3   3       6
4   4       9
Labels
3    1144
5    1116
8    1050
9    1049
1    1041
7    1022
2    1011
4     955
6     841
0     771
Name: count, dtype: int64


In [7]:
import torch
#from model import ResNet18CIFAR

device = 'cuda' if torch.cuda.is_available() else 'cpu'

# Load model
model = ResNet18CIFAR().to(device)
checkpoint = torch.load("best_model.pth", map_location=device)
model.load_state_dict(checkpoint)
model.eval()

print("Model loaded successfully!")


Model loaded successfully!


<ipython-input-7-7049ceb695fc>:8: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  checkpoint = torch.load("best_model.pth", map_location=device)


New run (after the 0.766 run) (march 7th 8 pm)
This model and traiing gave 0.88 on kaggle

In [16]:
import torch
import torch.nn as nn
import torch.nn.functional as F

# Squeeze-and-Excitation Block
class SEBlock(nn.Module):
    def __init__(self, channels, reduction=16):
        super(SEBlock, self).__init__()
        self.fc1 = nn.Linear(channels, channels // reduction)
        self.fc2 = nn.Linear(channels // reduction, channels)

    def forward(self, x):
        b, c, _, _ = x.size()
        squeeze = x.view(b, c, -1).mean(dim=2)
        excitation = F.relu(self.fc1(squeeze))
        excitation = torch.sigmoid(self.fc2(excitation)).view(b, c, 1, 1)
        return x * excitation

# Residual Block with SE and GroupNorm
class ResBlock(nn.Module):
    def __init__(self, in_channels, out_channels, stride=1):
        super(ResBlock, self).__init__()
        self.conv1 = nn.Conv2d(in_channels, out_channels, kernel_size=3, stride=stride, padding=1, bias=False)
        self.gn1 = nn.GroupNorm(8, out_channels)
        self.conv2 = nn.Conv2d(out_channels, out_channels, kernel_size=3, stride=1, padding=1, bias=False)
        self.gn2 = nn.GroupNorm(8, out_channels)
        self.se = SEBlock(out_channels)

        self.downsample = None
        if stride != 1 or in_channels != out_channels:
            self.downsample = nn.Sequential(
                nn.Conv2d(in_channels, out_channels, kernel_size=1, stride=stride, bias=False),
                nn.GroupNorm(8, out_channels)
            )

    def forward(self, x):
        identity = x
        out = self.conv1(x)
        out = F.relu(self.gn1(out))
        out = self.conv2(out)
        out = self.gn2(out)
        out = self.se(out)

        if self.downsample:
            identity = self.downsample(x)

        return F.relu(out + identity)

# Optimized ResNet-34 for CIFAR-10
class ResNet34CIFAR(nn.Module):
    def __init__(self, num_classes=10):
        super(ResNet34CIFAR, self).__init__()
        self.initial = nn.Sequential(
            nn.Conv2d(3, 64, kernel_size=3, stride=1, padding=1, bias=False),
            nn.GroupNorm(8, 64),
            nn.ReLU()
        )
        self.layer1 = self._make_layer(64, 64, 3, stride=1)
        self.layer2 = self._make_layer(64, 128, 4, stride=2)
        self.layer3 = self._make_layer(128, 256, 6, stride=2)
        self.layer4 = self._make_layer(256, 512, 3, stride=2)
        self.avgpool = nn.AdaptiveAvgPool2d(1)
        self.fc = nn.Linear(512, num_classes)

    def _make_layer(self, in_channels, out_channels, num_blocks, stride):
        layers = [ResBlock(in_channels, out_channels, stride)]
        for _ in range(1, num_blocks):
            layers.append(ResBlock(out_channels, out_channels, 1))
        return nn.Sequential(*layers)

    def forward(self, x):
        x = self.initial(x)
        x = self.layer1(x)
        x = self.layer2(x)
        x = self.layer3(x)
        x = self.layer4(x)
        x = self.avgpool(x)
        x = torch.flatten(x, 1)
        return self.fc(x)


In [20]:
import torch
import torch.nn as nn
import torch.optim as optim
import torchvision.transforms as transforms
import torchvision
from torch.utils.data import DataLoader, Dataset
import pickle
import numpy as np
from timm.optim.lookahead import Lookahead
#from model import ResNet34CIFAR  # Import optimized model
from PIL import Image
# Function to load CIFAR-10 from local files
def unpickle(file):
    with open(file, 'rb') as fo:
        data_dict = pickle.load(fo, encoding='bytes')
    return data_dict

# Custom dataset class
class CIFARDataset(Dataset):
    def __init__(self, data, labels, transform=None):
        self.data = data.reshape(-1, 3, 32, 32).astype("float32") / 255  # Normalize 0-1
        self.labels = labels
        self.transform = transform

    def __len__(self):
        return len(self.data)


    def __getitem__(self, idx):
        img = self.data[idx] * 255  # Convert float32 (0-1) to uint8 (0-255)
        img = img.astype("uint8")   # Ensure data type is uint8
        img = Image.fromarray(img.transpose(1, 2, 0))  # Convert to PIL image

        label = self.labels[idx]

        if self.transform:
            img = self.transform(img)  # Now the transform will work correctly

        return img, label

def load_cifar10_batches(root_dir):
    data_list, labels_list = [], []
    for i in range(1, 6):  # CIFAR-10 has 5 training batches
        batch_file = f"{root_dir}/data_batch_{i}"
        batch = unpickle(batch_file)
        data_list.append(batch[b'data'])
        labels_list.extend(batch[b'labels'])

    X_train = np.concatenate(data_list, axis=0)
    y_train = np.array(labels_list)
    return X_train, y_train

# Training function
def train():
    device = 'cuda' if torch.cuda.is_available() else 'mps' if torch.backends.mps.is_available() else 'cpu'
    print(f"Using device: {device}")

    # Load CIFAR-10 from local directory
    root_dir = "/content/cifar-10-python/cifar-10-batches-py"
    X_train, y_train = load_cifar10_batches(root_dir)

    # Data augmentation
    transform_train = transforms.Compose([
        transforms.RandomCrop(32, padding=4),
        transforms.RandomHorizontalFlip(),
        transforms.AutoAugment(),
        transforms.ToTensor(),
        transforms.Normalize((0.4914, 0.4822, 0.4465), (0.2470, 0.2435, 0.2616))
    ])

    train_dataset = CIFARDataset(X_train, y_train, transform=transform_train)
    train_loader = DataLoader(train_dataset, batch_size=512, shuffle=True, num_workers=4)

    # Initialize model
    model = ResNet34CIFAR().to(device)

    # Define loss and optimizer
    criterion = nn.CrossEntropyLoss(label_smoothing=0.1)
    base_optimizer = optim.AdamW(model.parameters(), lr=3e-4, weight_decay=5e-4)
    optimizer = Lookahead(base_optimizer)  # Lookahead wraps AdamW
    scheduler = optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=350)

    best_acc = 0.0

    # Training loop
    for epoch in range(350):
        model.train()
        running_loss = 0.0

        for images, labels in train_loader:
            images, labels = images.to(device), labels.to(device)
            optimizer.zero_grad()

            outputs = model(images)
            loss = criterion(outputs, labels)
            loss.backward()
            optimizer.step()

            running_loss += loss.item()

        scheduler.step()

        # Save best model
        if epoch % 50 == 0:
            torch.save(model.state_dict(), f"model_epoch_{epoch}.pth")

    torch.save(model.state_dict(), "best_model.pth")
    print("Training complete. Best model saved.")

if __name__ == "__main__":
    train()


Using device: cuda


Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7b26682c76a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/dist-packages/torch/utils/data/dataloader.py", line 1604, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.11/dist-packages/torch/utils/data/dataloader.py", line 1587, in _shutdown_workers
    if w.is_alive():
       ^^^^^^^^^^^^
  File "/usr/lib/python3.11/multiprocessing/process.py", line 160, in is_alive
    assert self._parent_pid == os.getpid(), 'can only test a child process'
           ^^^^^^^^^^^Exception ignored in: ^<function _MultiProcessingDataLoaderIter.__del__ at 0x7b26682c76a0>^
^Traceback (most recent call last):
^  File "/usr/local/lib/python3.11/dist-packages/torch/utils/data/dataloader.py", line 1604, in __del__
^    ^self._shutdown_workers()^^
^  File "/usr/local/lib/python3.11/dist-packages/torch/utils/data/dataloader.py", line 1587, in _shutdown_workers
^    ^if w.is_alive():^
^ ^  ^^ ^

Training complete. Best model saved.


In [22]:
import torch
import torchvision.transforms as transforms
import pandas as pd
import pickle
import os
import numpy as np
#from model import ResNet34CIFAR  # Import new model
from torch.utils.data import DataLoader, Dataset

def unpickle(file):
    with open(file, 'rb') as fo:
        data_dict = pickle.load(fo, encoding='bytes')
    return data_dict

class CIFARTestDataset(Dataset):
    def __init__(self, data, ids, transform=None):
        """
        data: numpy array of shape (N, 32, 32, 3)
        ids: array-like of test image IDs
        transform: transforms to apply
        """
        self.data = data
        self.ids = ids
        self.transform = transform

    def __len__(self):
        return len(self.data)

    def __getitem__(self, idx):
        img = self.data[idx].astype("uint8")
        img_id = self.ids[idx]
        if self.transform:
            img = self.transform(img)
        return img, img_id

def inference():
    device = 'cuda' if torch.cuda.is_available() else 'mps' if torch.backends.mps.is_available() else 'cpu'
    print(f"Using device: {device}")

    # Load optimized ResNet-34 model
    model = ResNet34CIFAR().to(device)
    model.load_state_dict(torch.load("/content/model_epoch_300.pth", map_location=device))
    model.eval()

    test_file = "/content/cifar_test_nolabel.pkl"
    test_dict = unpickle(test_file)
    print("Keys in test_dict:", test_dict.keys())

    test_images = test_dict[b'data'].reshape(-1, 32, 32, 3)  # Ensure correct shape
    test_ids = [str(i) for i in range(len(test_images))]  # Generate sequential IDs

    transform_test = transforms.Compose([
        transforms.ToPILImage(),
        transforms.ToTensor(),
        transforms.Normalize((0.4914, 0.4822, 0.4465), (0.2470, 0.2435, 0.2616))
    ])

    test_dataset = CIFARTestDataset(test_images, test_ids, transform=transform_test)
    test_loader = DataLoader(test_dataset, batch_size=128, shuffle=False, num_workers=4)

    predictions = []
    image_ids = []

    with torch.no_grad():
        for images, ids in test_loader:
            images = images.to(device)
            outputs = model(images)
            probabilities = torch.nn.functional.softmax(outputs, dim=1)  # Apply Softmax
            _, predicted = torch.max(probabilities, 1)
            predictions.extend(predicted.cpu().numpy().tolist())
            image_ids.extend(ids)

    submission = pd.DataFrame({"ID": image_ids, "Labels": predictions})
    submission.to_csv("submission.csv", index=False)
    print("Submission file saved as submission.csv")

if __name__ == "__main__":
    inference()


Using device: cuda
Keys in test_dict: dict_keys([b'data', b'ids'])


<ipython-input-22-a463baec993b>:42: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  model.load_state_dict(torch.load("/content/model_epoch_300.pth", map_location=device))


Submission file saved as submission.csv
